include/sys_clock.h

```diff
@@ -1,248 +1,109 @@
- * but if the HW timer cycles/sec, ticks/sec and ms/sec are all known
- * to be nicely related, then we can cheat with 32 bits instead.
- */
-
-#ifdef CONFIG_SYS_CLOCK_EXISTS
-
-#if defined(CONFIG_TIMER_READS_ITS_FREQUENCY_AT_RUNTIME) || \
-	(MSEC_PER_SEC % CONFIG_SYS_CLOCK_TICKS_PER_SEC) || \
-	(CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC % CONFIG_SYS_CLOCK_TICKS_PER_SEC)
-#define _NEED_PRECISE_TICK_MS_CONVERSION
-#endif
-
+#ifdef __cplusplus
+extern "C" {
 #endif
 
-static ALWAYS_INLINE s32_t z_ms_to_ticks(s32_t ms)
-{
-#ifdef CONFIG_SYS_CLOCK_EXISTS
-
-#ifdef _NEED_PRECISE_TICK_MS_CONVERSION
-	int cyc = sys_clock_hw_cycles_per_sec();
-
-	/* use 64-bit math to keep precision */
-	return (s32_t)ceiling_fraction((s64_t)ms * cyc,
-		((s64_t)MSEC_PER_SEC * cyc) / CONFIG_SYS_CLOCK_TICKS_PER_SEC);
+#ifdef CONFIG_SYS_TIMEOUT_64BIT
+typedef u64_t k_ticks_t;
 #else
-	/* simple division keeps precision */
-	s32_t ms_per_tick = MSEC_PER_SEC / CONFIG_SYS_CLOCK_TICKS_PER_SEC;
-
-	return (s32_t)ceiling_fraction(ms, ms_per_tick);
+typedef u32_t k_ticks_t;
 #endif
 
+#ifdef CONFIG_SYS_TIMEOUT_LEGACY_API
+/* Fallback API where timeouts are still 32 bit millisecond counts */
+typedef k_ticks_t k_timeout_t;
+#define K_TIMEOUT_TICKS(t) (t)
+#define K_TIMEOUT_GET(t) (t)
+#define K_TIMEOUT_EQ(a, b) ((a) == (b))
 #else
-	__ASSERT(ms == 0, "ms not zero");
-	return 0;
+/* New API going forward: k_timeout_t is an opaque struct */
+typedef struct { k_ticks_t ticks; } k_timeout_t;
+#define K_TIMEOUT_TICKS(t) ((k_timeout_t){ (k_ticks_t)t })
+#define K_TIMEOUT_GET(t) ((t).ticks)
+#define K_TIMEOUT_EQ(a, b) ((a).ticks == (b).ticks)
 #endif
-}
 
-static inline u64_t __ticks_to_ms(s64_t ticks)
-{
-#ifdef CONFIG_SYS_CLOCK_EXISTS
-	return (u64_t)ticks * MSEC_PER_SEC /
-	       (u64_t)CONFIG_SYS_CLOCK_TICKS_PER_SEC;
-#else
-	__ASSERT(ticks == 0, "ticks not zero");
-	return 0ULL;
-#endif
-}
+struct _timeout;
+typedef void (*_timeout_func_t)(struct _timeout *t);
 
-/*
- * These are only currently used by k_usleep(), but they are
- * defined here for parity with their ms analogs above. Note:
- * we don't bother trying the 32-bit intermediate shortcuts
- * possible with ms, because of the magnitudes involved.
- */
+struct _timeout {
+	sys_dnode_t node;
+	k_ticks_t dticks;
+	_timeout_func_t fn;
+};
 
-static inline s32_t z_us_to_ticks(s32_t us)
-{
-#ifdef CONFIG_SYS_CLOCK_EXISTS
-	return (s32_t) ceiling_fraction(
-		(s64_t)us * sys_clock_hw_cycles_per_sec(),
-		((s64_t)USEC_PER_SEC * sys_clock_hw_cycles_per_sec()) /
-		CONFIG_SYS_CLOCK_TICKS_PER_SEC);
+// FIXME: these are the new APIs, need doxygen.
+#ifdef CONFIG_SYS_TIMEOUT_64BIT
+#define K_TIMEOUT_MS(t) K_TIMEOUT_TICKS(k_ms_to_ticks_ceil64(t))
+#define K_TIMEOUT_US(t) K_TIMEOUT_TICKS(k_us_to_ticks_ceil64(t))
+#define K_TIMEOUT_CYC(t) K_TIMEOUT_TICKS(k_cyc_to_ticks_ceil64(t))
 #else
-	__ASSERT(us == 0, "us not zero");
-	return 0;
+#define K_TIMEOUT_MS(t) K_TIMEOUT_TICKS(k_ms_to_ticks_ceil32(t))
```

Comment:
Shouldn't `K_TIMEOUT_MS(x)` return `x` in legacy mode? since `k_sleep(5)` should behave as `k_sleep(K_TIMEOUT_MS(5))`. or not?